### PR TITLE
JDK-8310233: Fix THP detection on Linux

### PR DESCRIPTION
**Pull request by tstuefe**
Today, if we use UseTransparentHugePages, we assume that the *static* hugepage detection we do is valid for THPs:
- that THPs use the page size (in hotspot used as "default large page size") found in `/proc/memlimit` "Hugepagesize")
- that THPs are enabled if that page size is >0.

Both assumptions are incorrect:

- whether THPs are enabled should be checked at `/sys/kernel/mm/transparent_hugepage/enabled`, which is a tri-state value ("always", "madvise", "never"). THPs are available for the first two states.
- The page size employed by `khugepaged` is set in `/sys/kernel/mm/transparent_hugepage/hpage_pmd_size`. It can differ from the default page size used for static hugepages. For example, we could configure a system such that it uses 1G static hugepages, but the THP page size would still be 2M.

------

About the patch:

This is a limited, minimally invasive patch to fix THP detection. The patch aims to be easy to downport. There is more work to do, which I will do in subsequent RFEs.

Functionally, for *static* (non-THP) pages nothing changes. THP-mode now correctly detects THP support in the OS, and uses the correct page size (see examples below).

-------------

Example 1: System has THPs disabled, but static hugepages (1g, 2m) configured:

```
thomas@starfish $ cat /sys/kernel/mm/transparent_hugepage/enabled
always madvise [never]
thomas@starfish $ cat /proc/meminfo | grep Hugepage
Hugepagesize:    1048576 kB
```

Without patch, we incorrectly assume THPs are enabled, and that THP page size is 1G (!), which we then proceed and use as heap page size, causing the heap size to be rounded up from 512m -> 1G. But - even though it is printed as "1G page backed" in log output - in reality it will still 4K-page-backed: the `madvise(2)` we use to set the THP page size will be ignored by the system, since THPs are disabled.

```
thomas@starfish $ ./images/jdk/bin/java -Xmx512m -XX:+UseLargePages -XX:+UseTransparentHugePages -Xlog:pagesize -version
[0.001s][info][pagesize] Using the default large page size: 1G
[0.001s][info][pagesize] Usable page sizes: 4k, 2M, 1G
...
[0.016s][info][pagesize] Heap:  min=1G max=1G base=0x00000000c0000000 size=1G page_size=1G
```

With patch, we correctly refuse to use large pages (and we log more info):

```
thomas@starfish $ ./images/jdk/bin/java -Xmx512m -XX:+UseLargePages -XX:+UseTransparentHugePages -Xlog:pagesize -version
[0.001s][info][pagesize] Static hugepage support: 2M, 1G (default)
[0.001s][info][pagesize]   default pagesize: 1G
[0.001s][info][pagesize] Transparent hugepage (THP) support:
[0.001s][info][pagesize]   mode: never
[0.001s][warning][pagesize] UseLargePages disabled, no large pages configured and available on the system.
```

----------

Example 2: System has THPs enabled, but THP page size is just *2M*, whereas the system uses a static default hugepage size of *1G*:

```
thomas@starfish $ cat /sys/kernel/mm/transparent_hugepage/enabled
always [madvise] never
thomas@starfish $ cat /sys/kernel/mm/transparent_hugepage/hpage_pmd_size 
2097152
thomas@starfish $ cat /proc/meminfo | grep Hugepage
Hugepagesize:    1048576 kB
```

Without patch, THP page size is not correctly recognized as 2M. Instead, we again use 1G as page size for the heap:

```
thomas@starfish $ ./images/jdk/bin/java -Xmx512m -XX:+UseLargePages -XX:+UseTransparentHugePages -Xlog:pagesize -version
[0.001s][info][pagesize] Using the default large page size: 1G
[0.001s][info][pagesize] Usable page sizes: 4k, 2M, 1G
...
[0.010s][info][pagesize] Heap:  min=1G max=1G base=0x00000000c0000000 size=1G page_size=1G
```

With patch, we correctly identify the THP page size as 2M, and use that for the heap:

```
thomas@starfish $ ./images/jdk/bin/java -Xmx512m -XX:+UseLargePages -XX:+UseTransparentHugePages -Xlog:pagesize -version
[0.001s][info][pagesize] Static hugepage support: 2M, 1G (default)
[0.001s][info][pagesize]   default pagesize: 1G
[0.001s][info][pagesize] Transparent hugepage (THP) support:
[0.001s][info][pagesize]   mode: madvise
[0.001s][info][pagesize]   pagesize: 2M
[0.001s][info][pagesize] Large page support enabled. Usable page sizes: 4k, 2M
[0.001s][info][pagesize]  Default: 2M
...
[0.010s][info][pagesize] Heap:  min=8M max=512M base=0x00000000e0000000 size=512M page_size=2M
```
------------

Tests: GHAs all green. Local experiments on x64 Linux on machines with 1G pages succeeded.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8310233](https://bugs.openjdk.org/browse/JDK-8310233): Fix THP detection on Linux (**Bug** - P4)

### Reviewers
 * [Johan Sjölen](https://openjdk.org/census#jsjolen) (@jdksjolen - **Reviewer**) ⚠️ Review applies to [69c39c8a](https://git.openjdk.org/jdk/pull/14739/files/69c39c8aa4411bfd468cfc63faa7559728a10645)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14739/head:pull/14739` \
`$ git checkout pull/14739`

Update a local copy of the PR: \
`$ git checkout pull/14739` \
`$ git pull https://git.openjdk.org/jdk.git pull/14739/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14739`

View PR using the GUI difftool: \
`$ git pr show -t 14739`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14739.diff">https://git.openjdk.org/jdk/pull/14739.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14739#issuecomment-1616059461)

### CODE REVIEW

File: src/hotspot/os/linux/hugepages.cpp
Comment (jdksjolen) [inline src/hotspot/os/linux/hugepages.cpp:160]: As `print_on` prints several lines you want to use `NonInterleavingLogStream`.

Do this:

```c++
LogMessage(pagesize) lm;
NonInterleavingLogStream ls{LogLevelType::Info, lm};
if (ls.is_enabled()) {
  print_on(&ls);
}
```

Comment (jdksjolen) [reply src/hotspot/os/linux/hugepages.cpp:160]: I'm just using `{}`-init here, it's the same as using `()` so your choice on which to use.

**Review by jdksjolen — COMMENTED** [review]
Hi Thomas,

Thanks for these changes. I've some suggestions for the code, but the conceptual change is good.

File: src/hotspot/os/linux/hugepages.cpp
Comment (tstuefe) [reply src/hotspot/os/linux/hugepages.cpp:84]: Definitely simpler, but this code moved (almost) verbatim from the old `scan_default_large_page_size` in os_linux.cpp, and I would like to keep code changes to a minimum.

File: src/hotspot/os/linux/hugepages.cpp
Comment (tstuefe) [reply src/hotspot/os/linux/hugepages.cpp:65]: Again, unchanged code from before I'd like to keep unchanged to keep this patch minimally invasive. Mentally earmarked for future cleanups.

File: src/hotspot/os/linux/hugepages.cpp
Comment (tstuefe) [reply src/hotspot/os/linux/hugepages.cpp:93]: I'd like to leave it.

While I agree with you, Hotspot style guide says something like "write parentheses if order of ops is unclear". Most of the code in hotspot uses parentheses in assignments like this. I'd like to avoid mixing styles.

File: src/hotspot/os/linux/hugepages.cpp
Comment (tstuefe) [reply src/hotspot/os/linux/hugepages.cpp:68]: Absolutely agree, for a future cleanup, see above.

File: src/hotspot/os/linux/hugepages.cpp
Comment (tstuefe) [reply src/hotspot/os/linux/hugepages.cpp:160]: Okay. We are at initialization time, though; everything should still be single-threaded here. I dislike providing the log level twice, but LogTarget does not expose it.

Comment (tstuefe) [reply src/hotspot/os/linux/hugepages.cpp:160]: Wait, if I do this, I need to backport [8286180](https://bugs.openjdk.org/browse/JDK-8286180) before backporting this fix., right? :(

Oh well, chances are this does not integrate cleanly anyway.

Comment (tstuefe) [reply src/hotspot/os/linux/hugepages.cpp:160]: The more I think about this, the more doubts I have, sorry. Logging should not cost anything if disabled. But with your proposal, I pay for the construction of both LogMessage and NonInterleavingLogStream every time.

**Comment by tstuefe** [comment]
@jdksjolen Thanks a lot for your review!

See inline remarks. Most of your code suggestions are good, but this patch just moved the static hugepage detection parts out of os_linux.cpp, and left them (mostly) alone otherwise, to avoid adding regressions. I'll keep your input in mind for the next round of cleanups.

Wrt NonInterleavingLogStream, I decided to not do that. It is not needed, since we are single-threaded, and it without it I may still be lucky to get it integrated cleanly to at least 17.

I added a new repro case. Please take a look!

Thank you.

File: src/hotspot/os/linux/hugepages.cpp
Comment (jdksjolen) [reply src/hotspot/os/linux/hugepages.cpp:155]: >But with your proposal, I pay for the construction of both LogMessage and NonInterleavingLogStream every time.

Yeah, I made a mental footnote that `LogTarget` should be able to be used to initialize these inside of an if-block. UL is still a bit painful about these things. My suggestion sounds like a massive pain, I'm happy if you don't do it :).

Comment (tstuefe) [reply src/hotspot/os/linux/hugepages.cpp:155]: > > But with your proposal, I pay for the construction of both LogMessage and NonInterleavingLogStream every time.
> 
> Yeah, I made a mental footnote that `LogTarget` should be able to be used to initialize these inside of an if-block. UL is still a bit painful about these things. My suggestion sounds like a massive pain, I'm happy if you don't do it :).

Thank you @jdksjolen !

File: test/hotspot/jtreg/runtime/os/HugePageConfiguration.java
Comment (jdksjolen) [inline test/hotspot/jtreg/runtime/os/HugePageConfiguration.java:92]: Nit: You can use try-with-resources here.

File: test/hotspot/jtreg/runtime/os/HugePageConfiguration.java
Comment (jdksjolen) [inline test/hotspot/jtreg/runtime/os/HugePageConfiguration.java:128]: This leaks the file descriptor (never closed), use try-with-resources.

File: test/hotspot/jtreg/runtime/os/HugePageConfiguration.java
Comment (jdksjolen) [inline test/hotspot/jtreg/runtime/os/HugePageConfiguration.java:149]: This leaks the file descriptor (never closed), use try-with-resources.

**Review by jdksjolen — COMMENTED** [review]
Some of the methods seems to leak file descriptors, other than that tests looks good.

File: test/hotspot/jtreg/runtime/os/HugePageConfiguration.java
Comment (tstuefe) [reply test/hotspot/jtreg/runtime/os/HugePageConfiguration.java:92]: Makes sense

**Comment by tstuefe** [comment]
Thanks @jdksjolen! I changed the test accordingly.

**Review by jdksjolen — APPROVED** [review]
Thank you Thomas!

These changes look good to me now, I'm approving this.

**Comment by tstuefe** [comment]
Gentle ping, second review needed.

File: src/hotspot/os/linux/hugepages.cpp
Comment (tstuefe) [inline src/hotspot/os/linux/hugepages.cpp:51]: Reviewer notes: 
`scan_default_hugepagesize` is just a code move, it used to live in `scan_default_large_page_size` in os_linux.cpp. It is responsible for reading the default *static* hugepage size from /proc/meminfo. This code is mostly untouched to reduce chance for regressions (though the code could be tightened and cleaned for sure).

**Comment by tstuefe** [comment]
May I have a second review?

File: test/hotspot/jtreg/runtime/os/HugePageConfiguration.java
Comment (dholmes-ora) [inline test/hotspot/jtreg/runtime/os/HugePageConfiguration.java:65]: The `java.lang` is not necessary and typically not used.

File: src/hotspot/share/utilities/globalDefinitions.hpp
Comment (dholmes-ora) [inline src/hotspot/share/utilities/globalDefinitions.hpp:413]: Not sure this is worthy of inclusion in globalDefinitions. And I struggle to understand the use of "exact" in this context (pre-existing).

File: src/hotspot/os/linux/hugepages.cpp
Comment (dholmes-ora) [inline src/hotspot/os/linux/hugepages.cpp:92]: Why `bool` return when you don't check it? Also `*out == 0` also indicates failure. Though why use an out parameter instead of just returning the value?

**Review by dholmes-ora — COMMENTED** [review]
Not an area I am familiar with but scanning through the code this seems reasonable. A few minor comments below.

Thanks.

**Comment by tstuefe** [comment]
> Not an area I am familiar with but scanning through the code this seems reasonable. A few minor comments below.
> 
> Thanks.

Many thanks, @dholmes-ora! Remarks follow.

File: src/hotspot/os/linux/hugepages.cpp
Comment (tstuefe) [reply src/hotspot/os/linux/hugepages.cpp:92]: I will fix the caller to handle the error. Keeping error state from "0" return separate is just cleaner, and makes it possible to move/reuse this function should the need arise.

**Comment by tstuefe** [comment]
@dholmes-ora : I massaged in your feedback.

**Review by dholmes-ora — APPROVED** [review]
Nothing further from me.

Thanks.

**Comment by tstuefe** [comment]
Thanks, @dholmes-ora !

/integrate

**Pull request closed without merging**
